### PR TITLE
refactor(core): static-query schematic should handle binary expressions and function callbacks

### PR DESCRIPTION
**Pull request by devversion**
* See individual commits.

### CODE REVIEW

**Review by kara — APPROVED** [review]
LGTM, aside from duplicated test. Open question can be resolved in a follow-up.

File: packages/core/schematics/test/static_queries_migration_spec.ts [hunk 1]
```diff
@@ -639,6 +639,121 @@ describe('static-queries migration', () => {
...
+        }
+      `);
+
+      runMigration();
+
+      expect(tree.readContent('/index.ts'))
+          .toContain(`@${queryType}('test', { static: true }) query: any;`);
+    });
+
+    it('should detect callback with statically used queries', () => {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/test/static_queries_migration_spec.ts]: Good catch. :+1:

File: packages/core/schematics/test/static_queries_migration_spec.ts [hunk 2]
```diff
@@ -639,6 +639,121 @@ describe('static-queries migration', () => {
...
+            callSyncNested();
+          }
+        }
+      `);
+
+      runMigration();
+
+      expect(tree.readContent('/index.ts'))
+          .toContain(`@${queryType}('test', { static: true }) query: any;`);
+    });
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/test/static_queries_migration_spec.ts]: Hmm that is a good point. The initial thinking of the usage detection was that we just ignore the conditionals as we can't know to what the evaluate. So we always mark the query as `static` in that case.

Marking the query as _ambiguous_ with `static: true` sounds reasonable though. As you said, let's do/ think about this more in a follow-up. There are plenty other things that we discussed about. I added it to the query detection strategy hackmd.

**Comment by kara** [comment]
[presubmit](http://test/OCL:242510667:BASE:242510708:1554750575196:b31495f5)

**Comment by kara** [comment]
merge-assistance: I don't know what's going on with the "cla/google" check, but it's ready to go

**Pull request closed without merging**